Xamarin.Forms.Core/Registrar.cs

```diff
@@ -138,24 +119,76 @@ public Type GetHandlerTypeForObject(object obj)
 			return GetHandlerType(type);
 		}
 
-		bool LookupHandlerType(Type viewType, out Type handlerType)
+		bool LookupHandlerType(Type viewType, bool includeBaseTypes, out Type handlerType)
 		{
 			Type type = viewType;
 
-			while (type != null)
+			while (type != null && type != typeof(Element))
 			{
 				if (_handlers.ContainsKey(type))
 				{
 					handlerType = _handlers[type];
 					return true;
 				}
 
-				type = type.GetTypeInfo().BaseType;
+				if (includeBaseTypes)
+					type = type.GetTypeInfo().BaseType;
+				else
+					type = null;
 			}
 
 			handlerType = null;
 			return false;
 		}
+
+		void RegisterRenderWithTypes(Type viewType)
+		{
+			Type type = viewType;
+
+			// We're going to go through each type in this viewType's inheritance chain to look for classes
+			// decorated with a RenderWithAttribute. We're going to register each specific type with its
+			// renderer. 
+
+			while (type != null && type != typeof(Element))
+			{
+				// Only go through this process if we have not registered something for this type;
+				// we don't want RenderWith renderers to override ExportRenderers that are already registered.
+				// Plus, there's no need to do this again if we already have a renderer registered.
+				if (!LookupHandlerType(type, false, out Type specificTypeRenderer))
```

Comment:
reduce nesting by reversing the `if`

```csharp
if (Lookup(...))
  break;
```

Reply:
(this is the reason why I recommended a do-while)

Reply:
I still need to go through the base types, since I'm going top-down. I can't break here.